python/TestHarness/testers/bench.py

```diff
@@ -0,0 +1,330 @@
+          realtime_secs REAL
+        );'''
+
+        CREATE_PERFLOG_TABLE = '''CREATE TABLE IF NOT EXISTS perflog
+        (
+          benchmark_id INTEGER,
+          run INTEGER,
+          field TEXT,
+          subfield TEXT,
+          exec_count INTEGER,
+          self_time_secs REAL,
+          cum_time_secs REAL
+        );'''
+
+        self.fname = fname
+        self.conn = sqlite3.connect(fname)
+        c = self.conn.cursor()
+        c.execute(CREATE_BENCH_TABLE)
+        c.execute(CREATE_TIMES_TABLE)
+        c.execute(CREATE_PERFLOG_TABLE)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        self.close()
+
+    def revisions(self):
+        c = self.conn.cursor()
+        c.execute('SELECT revision,date FROM benchmarks GROUP BY revision ORDER BY date ASC')
+        rows = c.fetchall()
+        revs = []
+        times = []
+        for r in rows:
+            revs.append(r[0])
+            times.append(r[1])
+        return revs, times
+
+    def list(self, revision, benchmark=''):
+        c = self.conn.cursor()
+        if benchmark == '':
+            c.execute('SELECT id,name,executable,input_file FROM benchmarks WHERE INSTR(revision,?) ORDER BY date ASC', (revision,))
+        else:
+            c.execute('SELECT id,name,executable,input_file FROM benchmarks WHERE INSTR(revision,?) AND INSTR(name,?) ORDER BY date ASC', (revision,benchmark))
+        benchmarks = c.fetchall()
+        return benchmarks
+
+    def load_times(self, bench_id):
+        c = self.conn.cursor()
+        c.execute('SELECT realtime_secs FROM timings WHERE benchmark_id=?', (bench_id,))
+        ents = c.fetchall()
+        real = []
+        for ent in ents:
+            real.append(float(ent[0]))
+        return real
+
+    def load(self, revision, bench_name):
+        """loads and returns a Bench object for the given revision and benchmark name"""
+        entries = self.list(revision, benchmark=bench_name)
+        b = entries[0]
+        real = self.load_times(b[0])
+        return Bench(b[1], test=Test(b[2], b[3]), realruns=real)
+
+    def store(self, benchmark, rev=None):
+        """stores a (run/executed) Bench in the database. if rev is None, git revision is retrieved from git"""
+        ex = benchmark.test.executable
+        (ex_name, ex_method) = os.path.basename(ex).rsplit('-', 1)
+        infile = benchmark.test.infile
+        timestamp = time.time()
+        date = timestamp
+        if rev is None:
+            rev, date = git_revision()
+        load = os.getloadavg()[0]
+
+        c = self.conn.cursor()
+        c.execute('INSERT INTO benchmarks (name,executable,executable_name,executable_method,input_file,timestamp,revision,date,load) VALUES (?,?,?,?,?,?,?,?,?)',
+                (benchmark.name, ex, ex_name, ex_method, infile, timestamp, rev, date, load))
+        self.conn.commit()
+
+        c.execute('SELECT max(id) from benchmarks')
+        bench_id = int(c.fetchone()[0])
```

Comment:
This isn't exactly be practice for obtaining the primary key from an insert. There's usually a more robust way. Have you checked that out?

Reply:
Yes - I just hacked this in quick - I'll make it better.